flink-core/src/test/java/org/apache/flink/api/common/state/StateDescriptorTest.java

```diff
@@ -200,6 +204,31 @@ public void testEqualsSameNameAndTypeDifferentClass() throws Exception {
 		assertNotEquals(descr1, descr2);
 	}
 
+	@Test
+	public void testSerializerLazyInitializeInParallel() throws Exception {
+		final String name = "testSerializerLazyInitializeInParallel";
+		// use PojoTypeInfo which will create a new serializer when createSerializer is invoked.
+		final TestStateDescriptor<String> desc =
+			new TestStateDescriptor<>(name, new PojoTypeInfo<>(String.class, new ArrayList<>()));
+		final int threadNumber = 20;
+		final ArrayList<Thread> threads = new ArrayList<>(threadNumber);
+		final ExecutionConfig executionConfig = Mockito.mock(ExecutionConfig.class);
+		final ConcurrentHashMap<Integer, TypeSerializer<String>> serializers = new ConcurrentHashMap<>();
+		for (int i = 0; i < threadNumber; i++) {
+			threads.add(new Thread(() -> {
+				desc.initializeSerializerUnlessSet(executionConfig);
+				TypeSerializer<String> serializer = desc.getOriginalSerializer();
+				serializers.put(System.identityHashCode(serializer), serializer);
+			}));
+		}
+		threads.forEach(Thread::start);
+		for (Thread t : threads) {
+			t.join();
```

Comment:
We should maybe also check for exceptions in these threads. You can use Flink's `CheckedThread` for this purpose.